### ISSUE TITLE
Jump in acc/CER at end of epoch

### ISSUE THREAD

**Issue by jan-schuchardt** [espnet/espnet#959]
Hi,

I observed the following phenomenon when looking at the train.log file after training RNN models on Tedlium2:

Between the last iteration of epoch n and the 100th iteration of epoch n+1, there's often a noticeable increase increase in accuracy (often > 0.5% and definitely larger than intra-epoch improvements), even if the accuracy appears to have converged beforehand.

Is there some part of the training algorithm that would explain this?
Otherwise, I think there might be some data-leakage problem, e.g. updates being performed on the validation set when calculating validation accuracy.

**Comment by sw005320** [comment]
This may be due to the epsilon decay. Can you paste the learning curve (acc.png or loss.png generated under `results/` directory or the snapshot of the tensorboard?

**Comment by jan-schuchardt** [comment]
Hi, sorry for the late reply.
I ran the training without tensorboard logging (and before tensorboard logging was performed every 100 iterations) and the acc.png / loss.png plots are not informative, as they only plot per-epoch accuracy.
So I just created plots from the "results/log" file. Orange ticks indicate the end of an epoch.

The problem only seems to become apparent in low-capacity models:

Here is a very extreme example (Note: This is with a custom encoder layer, that applies a frame-wise transformations, followed by 1d convolutions along the time-dimension. I didn't change anything about the training algorithm though):

![rnn_custom](https://user-images.githubusercontent.com/13921292/61620491-31e6c500-ac71-11e9-955f-0737fd96089c.png)

The same the same pattern (though less pronounced) can be observed default blstmp encoder with only a single layer and 320 units (Look at iteration 24000 to 38000 and 53000 to 63000)

![rnn_1_layer](https://user-images.githubusercontent.com/13921292/61620704-a7529580-ac71-11e9-9ea7-a2f993364a26.png)

**Comment by sw005320** [comment]
Thanks for your detailed analysis. I think such jumps may be reasonable because the initial iterations after every epoch use the same training data appeared in the previous epoch and could have some discontinuous accuracy jumps compared with other iterations. But I'm not sure.
